From: наб <nabijaczleweli@nabijaczleweli.xyz>
Date: Wed, 14 Dec 2022 17:52:04 +0100
Subject: Update manpages

man: Fix getopts documentation

  The explicit arguments were missing, also exchange expr subst for
  arithmetic and fix the spacing around Bell Labs

man: Document jobs builtin

man: Note chdir being the same builtin as cd

man: Document kill builtin

The manual now contains all built-ins

man: Reword to avoid confusion v/v printf Ar argument[s]/arguments

  The current wording says that given
    printf a b c d
  a is the format, c and d are processed as noted,
  but b is unspecified

man: printf 'X, X is a byte under dash

  Multiple issues:
    * the encoding is not always ASCII
    * what ASCII code is assigned to я
    * dash isn't internationalised
      (this is nonconformant but out of scope),
      and uses the next /byte/; in a UTF-8 locale:
        $ printf %d\\n \'ą
        196
        $ printf %d\\n \'я
        196
      this is in contrast to POSIX (and bash), which says:
        > If the leading character is a single-quote or double-quote,
        > the value shall be the numeric value in the underlying codeset
        > of the character following the single-quote or double-quote.
      (i.e. mbrtowc(&val, argv[n], ...))

man: Document false built-in

  Only true was documented, add false just below it
  (out of order, but so is true, and the grouping makes much more sense).

man: Document ulimit -w

Bug-Debian: https://bugs.debian.org/558607
Bug-Debian: https://bugs.debian.org/975325
---
 src/dash.1    | 97 +++++++++++++++++++++++++++++++++++++++++++----------------
 src/options.c |  2 +-
 2 files changed, 72 insertions(+), 27 deletions(-)

diff --git a/src/dash.1 b/src/dash.1
index 4bb348a..39cec2f 100644
--- a/src/dash.1
+++ b/src/dash.1
@@ -1095,6 +1095,8 @@ etc).
 .It :
 .It true
 A null command that returns a 0 (true) exit value.
+.It false
+A null command that returns a 1 (false) exit value.
 .It \&. file
 The commands in the specified file are read and executed by the shell.
 .It alias Op Ar name Ns Op Ar "=string ..."
@@ -1143,8 +1145,8 @@ Do not execute the command but
 search for the command and print the absolute pathname
 of utilities, the name for builtins or the expansion of aliases.
 .El
-.It cd Ar -
-.It Xo cd Op Fl LP
+.It cd|chdir Ar -
+.It Xo cd|chdir Op Fl LP
 .Op Ar directory
 .Xc
 Switch to the specified directory (default
@@ -1344,13 +1346,12 @@ The number of previous commands that are accessible.
 .El
 .It fg Op Ar job
 Move the specified job or the current job to the foreground.
-.It getopts Ar optstring var
+.It getopts Ar optstring var Op Ar arg ...
 The
 .Tn POSIX
 .Ic getopts
 command, not to be confused with the
-.Em Bell Labs
--derived
+.Em Bell Labs Ns -derived
 .Xr getopt 1 .
 .Pp
 The first argument should be a series of letters, each of which may be
@@ -1388,6 +1389,12 @@ then
 .Ev OPTARG
 will be unset.
 .Pp
+By default, the variables
+.Va $1 , ... , $n
+are inspected; if
+.Ar arg Ns s
+are specified, they'll be parsed instead.
+.Pp
 .Va optstring
 is a string of recognized option letters (see
 .Xr getopt 3 ) .
@@ -1432,7 +1439,7 @@ do
 	\\?)	echo $USAGE; exit 1;;
 	esac
 done
-shift `expr $OPTIND - 1`
+shift $((OPTIND - 1))
 .Ed
 .Pp
 This code will accept any of the following as equivalent:
@@ -1464,6 +1471,47 @@ The
 .Fl r
 option causes the hash command to delete all the entries in the hash table
 except for functions.
+.It jobs Oo Fl lp Oc Op Ar job ...
+Display the status of all, or just the specified,
+.Ar job Ns s :
+.Bl -tag -compact -offset 5n -width "By default"
+.It By default
+display the job number, currency
+.Pq Sy +-
+status, if any, the job state, and its shell command.
+.It Fl l
+also output the PID of the group leader, and just the PID and shell commands
+of other members of the job.
+.It Fl p
+Display only leader PIDs, one per line.
+.El
+.It kill Oo Fl s Ar sigspec | Fl Ns Ar signum | Fl Ns Ar sigspec Oc Op Ar pid | job ...
+Equivalent to
+.Xr kill 1 ,
+but a
+.Ar job
+spec may also be specified.
+Signals can be either case-insensitive names without
+.Dv SIG
+prefixes or decimal numbers; the default is
+.Dv TERM .
+.It kill Fl l Op Ar signum | exitstatus
+List available signal names without the
+.Dv SIG
+prefix
+.Pq Ar sigspec Ns s .
+If
+.Ar signum
+specified, display just the
+.Ar sigspec
+for that signal.
+If
+.Ar exitstatus
+specified
+.Pq > Sy 128 ,
+display just the
+.Ar sigspec
+that caused it.
 .It pwd Op Fl LP
 builtin command remembers what the current directory
 is rather than recomputing it each time.
@@ -1528,41 +1576,36 @@ With the
 option specified the output will be formatted suitably for non-interactive use.
 .\".Pp
 .It Xo printf Ar format
-.Op Ar arguments  ...
+.Oo Ar value Oc Ns ...
 .Xc
 .Ic printf
-formats and prints its arguments, after the first, under control
-of the
-.Ar format  .
-The
-.Ar format
-is a character string which contains three types of objects: plain characters,
+formats and prints its arguments according to
+.Ar format ,
+a character string which contains three types of objects: plain characters,
 which are simply copied to standard output, character escape sequences which
 are converted and copied to the standard output, and format specifications,
 each of which causes printing of the next successive
-.Ar argument  .
+.Ar value .
 .Pp
-The
-.Ar arguments
-after the first are treated as strings if the corresponding format is
+Each
+.Ar value
+is treated as a string if the corresponding format specification is
 either
 .Cm b ,
-.Cm c
+.Cm c ,
 or
 .Cm s ;
-otherwise it is evaluated as a C constant, with the following extensions:
-.Pp
+otherwise it is evaluated as a C constant, with the following additions:
 .Bl -bullet -offset indent -compact
 .It
 A leading plus or minus sign is allowed.
 .It
-If the leading character is a single or double quote, the value is the
-.Tn ASCII
-code of the next character.
+If the leading character is a single or double quote, the value of the next byte.
 .El
 .Pp
-The format string is reused as often as necessary to satisfy the
-.Ar arguments  .
+The format string is reused as often as necessary until all
+.Ar value Ns s
+are consumed.
 Any extra format specifications are evaluated with zero or the null
 string.
 .Pp
@@ -2123,7 +2166,7 @@ printed; for commands and tracked aliases the complete pathname of the
 command is printed.
 .It ulimit Xo
 .Op Fl H \*(Ba Fl S
-.Op Fl a \*(Ba Fl tfdscmlpnv Op Ar value
+.Op Fl a \*(Ba Fl tfdscmlpnvwr Op Ar value
 .Xc
 Inquire about or set the hard or soft limits on processes or set new
 limits.
@@ -2176,6 +2219,8 @@ show or set the limit on the number files a process can have open at once
 .It Fl v
 show or set the limit on the total virtual memory that can be
 in use by a process (in kilobytes)
+.It Fl w
+show or set the limit on the total number of locks held by a process
 .It Fl r
 show or set the limit on the real-time scheduling priority of a process
 .El
diff --git a/src/options.c b/src/options.c
index a46c23b..3158498 100644
--- a/src/options.c
+++ b/src/options.c
@@ -410,7 +410,7 @@ getoptscmd(int argc, char **argv)
 	char **optbase;
 
 	if (argc < 3)
-		sh_error("Usage: getopts optstring var [arg]");
+		sh_error("Usage: getopts optstring var [arg...]");
 	else if (argc == 3) {
 		optbase = shellparam.p;
 		if ((unsigned)shellparam.optind > shellparam.nparam + 1) {
